File: solutions/c/dnd-character/1/dnd_character.c
#include "dnd_character.h"
#include <stdlib.h>
#include <time.h>


int modifier(int score)
{
    int mod = score - 10;
    if(mod % 2 < 0)
        mod--;
    return mod / 2;
}

int ability(void)
{
    static int initialized = 0;
    if(!initialized)
    {
        srand(time(NULL));
        initialized = 1;
    }
    int sum = 0, smallest = 6;
    for(int i = 0; i < 4; i++)
    {
        int roll = rand() % (6) + 1;
        sum += roll;
        if(roll < smallest)
            smallest = roll;
    }
    return sum - smallest;
}

dnd_character_t make_dnd_character(void)
{
    dnd_character_t character;
    int *abilities[] = {
        &character.strength,
        &character.dexterity,
        &character.constitution,
        &character.intelligence,
        &character.wisdom,
        &character.charisma
    };
    
    for(size_t i = 0; i < sizeof(abilities) / sizeof(*abilities); i++)
        *abilities[i] = ability();
    
    character.hitpoints = modifier(character.constitution) + 10;
    return character;
}

File: solutions/java/dnd-character/1/src/main/java/DnDCharacter.java
import java.util.List;
import java.util.ArrayList;
import java.util.concurrent.ThreadLocalRandom;

class DnDCharacter
{
    private int strength;
    private int dexterity;
    private int constitution;
    private int intelligence;
    private int wisdom;
    private int charisma;
    private int modifier;

    public DnDCharacter()
    {
        this.strength = ability(rollDice());
        this.dexterity = ability(rollDice());
        this.constitution = ability(rollDice());
        this.intelligence = ability(rollDice());
        this.wisdom = ability(rollDice());
        this.charisma = ability(rollDice());
    }
    
    int ability(List<Integer> scores)
    {
        int sum = 0;
        int min = 6;
        for (int i = 0; i < 4; i++)
        {
            int currentScore = scores.get(i);
            sum += currentScore;
            if (currentScore < min)
            {
                min = currentScore;
            }
        }
        return sum - min;
    }

    List<Integer> rollDice()
    {
        List<Integer> scores = new ArrayList<>();
        for (int i = 0; i < 4; i++)
        {
            scores.add(ThreadLocalRandom.current().nextInt(1, 7));    
        }
        return scores;
    }

    int modifier(int input)
    {
        double mod = input - 10;
        mod /= 2;
        return (int)Math.floor(mod);
    }

    int getStrength()
    {
        return this.strength;
    }

    int getDexterity()
    {
        return this.dexterity;
    }

    int getConstitution()
    {
        return this.constitution;
    }

    int getIntelligence()
    {
        return this.intelligence;
    }

    int getWisdom()
    {
        return this.wisdom;
    }

    int getCharisma()
    {
        return this.charisma;
    }

    int getHitpoints()
    {   
        int mod = modifier(this.constitution);
        return mod + 10;
    }
}
